core/services/head_tracker.go

```diff
@@ -247,22 +250,122 @@ func (ht *HeadTracker) listenForNewHeads() {
 	defer ht.listenForNewHeadsWg.Done()
 	defer func() {
 		err := ht.unsubscribeFromHead()
-		logger.ErrorIf(err, "failed when unsubscribe from head")
+		logger.WarnIf(errors.Wrap(err, "HeadTracker failed when unsubscribe from head"))
 	}()
 
 	for {
 		if !ht.subscribe() {
 			return
 		}
 		if err := ht.receiveHeaders(); err != nil {
-			logger.Errorw(fmt.Sprintf("Error in new head subscription, unsubscribed: %s", err.Error()), "err", err)
+			logger.Errorw(fmt.Sprintf("HeadTracker: Error in new head subscription, unsubscribed: %s", err.Error()), "err", err)
 			continue
 		} else {
 			return
 		}
 	}
 }
 
+func (ht *HeadTracker) backfiller() {
+	defer ht.listenForNewHeadsWg.Done()
+	for {
+		select {
+		case <-ht.done:
+			return
+		case <-ht.backfillMB.Notify():
+			for {
+				head := ht.backfillMB.Retrieve()
+				if head == nil {
+					break
+				}
+				h, is := head.(models.Head)
+				if !is {
+					logger.Errorf("HeadTracker: invariant violation, expected %T but got %T", models.Head{}, head)
+					continue
+				}
+				ctx, cancel := utils.ContextFromChan(ht.done)
+				err := ht.Backfill(ctx, h, ht.store.Config.EthFinalityDepth())
+				cancel()
+				if err != nil {
+					logger.Warnw("HeadTracker: unexpected error while backfilling heads", "err", err)
+				}
+			}
+		}
+	}
+}
+
+// Backfill given a head will fill in any missing heads up to the given depth
+func (ht *HeadTracker) Backfill(ctx context.Context, head models.Head, depth uint) (err error) {
+	if uint(head.ChainLength()) >= depth {
+		return nil
+	}
+	baseHeight := head.Number - int64(depth-1)
+	if baseHeight < 0 {
+		baseHeight = 0
+	}
+
+	return ht.backfill(ctx, head.EarliestInChain(), baseHeight)
+}
+
+// backfill fetches all missing heads up until the base height
+func (ht *HeadTracker) backfill(ctx context.Context, head models.Head, baseHeight int64) (err error) {
+	if head.Number <= baseHeight {
+		return nil
+	}
+	mark := time.Now()
+	fetched := 0
+	logger.Debugw("HeadTracker: starting backfill",
+		"blockNumber", head.Number,
+		"id", "head_tracker",
+		"n", head.Number-baseHeight,
+		"fromBlockHeight", baseHeight,
+		"toBlockHeight", head.Number-1)
+	defer func() {
+		logger.Debugw("HeadTracker: finished backfill",
+			"fetched", fetched,
+			"blockNumber", head.Number,
+			"time", time.Since(mark),
+			"id", "head_tracker",
+			"n", head.Number-baseHeight,
+			"fromBlockHeight", baseHeight,
+			"toBlockHeight", head.Number-1,
+			"err", err)
+	}()
+
+	for i := head.Number - 1; i >= baseHeight; i-- {
+		// NOTE: Sequential requests here mean it's a potential performance bottleneck, be aware!
+		var existingHead *models.Head
+		existingHead, err = ht.store.HeadByHash(head.ParentHash)
+		if err != nil {
+			return errors.Wrap(err, "HeadByHash failed")
+		}
+		if existingHead != nil {
+			head = *existingHead
+			continue
+		}
+		head, err = ht.fetchAndSaveHead(ctx, i)
```

Comment:
worth making a `fetchAndSaveHeads(ctx, missingHeads)` which uses `BatchCallContext` ?

Reply:
Ah, there is a reason we don't do this - because we don't know what heads we need to fetch. We can reduce the total load on the eth node by fetching in series. Since this is now a background task it doesn't matter if this takes a bit longer.